src/Raven.Client/Documents/AI/AiConversation.cs

```diff
@@ -66,35 +74,91 @@ public void SetUserPrompt(string userPrompt)
-    public string Id => _conversationId ?? throw new InvalidOperationException($"This is a new conversation, the ID wasn't set yet, you have to call {nameof(Run)}/{nameof(RunAsync)}");
-    public AiConversationResult Run() => AsyncHelpers.RunSync(() => RunAsync());
+    public void Handle<TArgs>(string actionName, Func<TArgs, Task<object>> action, AiHandleErrorStrategy errorStrategy = AiHandleErrorStrategy.Default)
+        where TArgs : class
+    {
+        var t = new AiActionContext<TArgs>(_aiOperations, action, errorStrategy);
+        AddAction(actionName, t);
+    }
 
-    public async Task<AiConversationResult> RunAsync(CancellationToken token = default)
+    public void Handle<TArgs>(string actionName, Func<TArgs, object> action, AiHandleErrorStrategy errorStrategy = AiHandleErrorStrategy.Default) where TArgs : class
     {
-        IMaintenanceOperation<ConversationResult<T>> op;
-        if (string.IsNullOrWhiteSpace(_conversationId))
+        var t = new AiActionContext<TArgs>(_aiOperations, action, errorStrategy);
+        AddAction(actionName, t);
+    }
+
+    private void AddAction(string actionName, IAiActionContext t)
+    {
+        if (_invocations.ContainsKey(actionName))
+            throw new InvalidOperationException($"Action '{actionName}' already exists.");
+
+        _invocations.Add(actionName, t);
+    }
+  
+    public AiAnswer<TAnswer> Run<TAnswer>() => AsyncHelpers.RunSync(() => RunAsync<TAnswer>());
+
+    public async Task<AiAnswer<TAnswer>> RunAsync<TAnswer>(CancellationToken token = default)
+    {
+        while (true)
         {
-            op = new RunConversationOperation<T>(_agentId, _userPrompt, _parameters);
+            var r = await RunAsyncInternal<TAnswer>(token).ConfigureAwait(false);
+            if (r.Status == AiConversationResult.ActionRequired)
+            {
+                if (_actionRequests.Count == 0)
+                    throw new InvalidOperationException("There are no action requests to process.");
+
+                using (var ctx = JsonOperationContext.ShortTermSingleUse())
+                {
+                    for (int i = _actionRequests.Count - 1; i >= 0; i--)
+                    {
+                        var action = _actionRequests[i];
+                        if (_invocations.TryGetValue(action.Name, out var invocation))
+                        {
+                            {
+                                var response = await invocation.ExecuteAsync(ctx, action.Arguments, token).ConfigureAwait(false);
+                                AddActionResponse(action.ToolId, response);
+                            }
+
+                            _actionRequests.RemoveAt(i);
+                        }
+                    }
+                }
+
+                // we responded to all the action requests, so we can continue
+                if (_actionRequests.Count == 0)
```

Comment:
That is wrong. You _always_ need to submit whatever you have to the database, then get the _remainder_.
Otherwise, you may execute actions twice, or _not_ submit items to the database to close the tools, because you closed all of them